### ISSUE TITLE
Can't use const in a for-of loop

### ISSUE THREAD

**Issue by ivan** [facebook/flow#468]
With this `constfail.js`:

```js
"use strict";

const ok = 3;

for(const x of [1, 2]) {
        console.log(x);
}
```

flow breaks on the `const` in the for-of loop:

```
# cat constfail.js | flow check-contents

-:5:5,9: Unexpected token const

Found 1 error
```

**Comment by ivan** [comment]
Oh, never mind, I see that const and let aren't supported anywhere yet.

**Comment by samwgoldman** [comment]
@ivan I'm working on it, but are you sure that const is OK as a loop variable? I can see it both ways: On one hand, the variable is "re-bound" each iteration, so it's like getting a fresh `x` each time. On the other hand, `x` has a different value each iteration, and is thus non-constant.

Can you point to any evidence that `const` is valid here? Thanks!

**Comment by ivan** [comment]
Aha, you're right that the `const` in my loop should break at runtime.  I was using V8 4.2 which incorrectly allowed it, but it was fixed a short time ago in https://codereview.chromium.org/977543002/diff/1/src/parser.cc

But, as the tests in https://codereview.chromium.org/977543002/diff/1/test/mjsunit/harmony/block-const-assign.js show, `const` is still allowed here as long as the variable isn't reassigned in a subsequent loop iteration.  So a type checker should allow const here, unless it can prove that the `const` variable is reassigned.

**Comment by ivan** [comment]
Oops.  As mentioned in https://code.google.com/p/v8/issues/detail?id=4144#c1, `const` is always allowed in a `for-of` loop because the variables are re-bound in each iteration.

**Comment by samwgoldman** [comment]
@ivan Thanks for the resources! I'll make sure to test that functionality against my implementation.

**Comment by simonwelsh** [comment]
By the looks of #337, it won't even with `const` support, since flow doesn't currently support `for-of`.

**Comment by samwgoldman** [comment]
Well, parsing support landed in 280678da1f07588f8ce3f85765e3e2e7e1533d19. Not much left to do for for-or.
